### PR TITLE
Schedule page

### LINKED ISSUE #23: Move Titles to NavBar on Mobile View
Move the page headers to the NavBar in the mobile view, leave as is on desktop

### PR DESCRIPTION
**Pull request by thomasp18**
Implementation of the schedule function/page

### CODE REVIEW

File: next-app/app/schedule/page.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                                                        <td>{mval.teamA.join(', ')}</td>
+                                                        <td>{mval.ascore}</td>
+                                                        <td>{mval.bscore}</td>
+                                                        <td>{mval.teamB.join(', ')}</td>
+                                                    </tr>
+                                                );
+                                            }
+                                        })}
+                                    </tbody>
+                                </table>
```
Comment (shubunchu) [inline next-app/app/schedule/page.js]: for dates that don't contain any matches, I think it would look nicer if the table were replaced with a paragraph (e.g., "no matches found"), rather than an empty table

> image

File: next-app/app/page.js [hunk 2]
```diff
@@ -138,41 +197,94 @@ function WeatherData({ apiData }) {
...
+                                    The date &apos;<b>{apiDate}</b>&apos; was added to the schedule!
+                                  </p>
+                                ) : (
+                                  <p>
+                                    <b>{apiDate}</b> is non-unique! Choose a different date.
+                                  </p>
+                                )}
+                                <Button href="/schedule">
+                                  Go to schedule <i className="bi bi-caret-right-fill"></i>
+                                </Button>
```
Comment (shubunchu) [inline next-app/app/page.js]: maybe we can add an undo button here?

> image

currently, there isn't a way to unschedule from the home page. maybe we should add a confirmation modal, or simply have the "Schedule Date" button turn into a "Unschedule Date" button when clicked? the non-unique toast seems a little odd in this workflow

File: next-app/app/components/navbar/navbar.js [hunk 3]
```diff
@@ -1,34 +1,33 @@
...
-  );
+    let page = {
+        '/': 'Home', '/schedule': 'Schedule', '/score': 'Scorekeeper', '/matches': 'Match History', '/players': 'Players'
+    };
+
+    return (
+        <Navbar expand="lg" className="bg-body-tertiary">
+            <Container>
+                <Navbar.Brand href="/">
+                    <Image alt='' src='./favicon.ico' style={{ borderRadius: 7 }}></Image>
```
Comment (shubunchu) [inline next-app/app/components/navbar/navbar.js]: 🐵❓

File: next-app/app/page.js [hunk 4]
```diff
@@ -24,7 +84,14 @@ export default function Home() {
       <h1 className="display-1 text-center p-auto m-auto pt-2">PikoWeatherer</h1>
       <br />
       <div className="text-center">
-        <WeatherData apiData={response} />
+        <WeatherData
+          weatherData={weather}
+          addDate={addDate}
+          show={show}
```
Comment (shubunchu) [inline next-app/app/page.js]: use a more descriptive name, like `showToast`

File: next-app/app/page.js [hunk 5]
```diff
@@ -138,41 +197,94 @@ function WeatherData({ apiData }) {
...
+                              <b>{value.precipitation}%</b>
+                            </div>
+                          </ListGroup.Item>
+                          <ListGroup.Item className="card-text">
+                            <div className="float-start">Forecast: </div>
+                            <div className="float-end">
+                              <b>{value.weathercode}</b>
+                            </div>
+                          </ListGroup.Item>
+                        </ListGroup>
```
Comment (shubunchu) [inline next-app/app/page.js]: maybe omit the colon delimiters, now that the values are right-aligned? just a stylistic choice, but I think it looks nicer (see Forecast)
> image

File: next-app/app/page.js [hunk 6]
```diff
@@ -138,41 +197,94 @@ function WeatherData({ apiData }) {
...
+                                </strong>
+                              </Toast.Header>
+                              <Toast.Body>
+                                {dateRequested === 'success' ? (
+                                  <p>
+                                    The date &apos;<b>{apiDate}</b>&apos; was added to the schedule!
+                                  </p>
+                                ) : (
+                                  <p>
+                                    <b>{apiDate}</b> is non-unique! Choose a different date.
```
Comment (shubunchu) [inline next-app/app/page.js]: I think we can reword this to something along the lines of "this date is already scheduled"

File: next-app/app/schedule/page.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+                                <div className='text-end'>
+                                    <Button variant='danger' size="lg" onClick={unscheduleID}><i className='bi bi-trash-fill'></i></Button>
```
Comment (shubunchu) [inline next-app/app/schedule/page.js]: I think we should use a small (or regular) button size here, large takes up a lot of space - almost the same height as a game entry

large (current):
> image

small:
> image

and maybe a confirmation modal/"are you sure" prompt on the button? I feel like it's warranted given it's a destructive action

**Review by shubunchu — COMMENTED** [review]
added a few suggestions 🐒

**Comment by KyungjinJPark** [comment]
is it possible to make the toast transparent?

**Comment by KyungjinJPark** [comment]
schedule page score to one column

File: next-app/app/api/matches/route.js [hunk 8]
```diff
@@ -34,3 +34,28 @@ export async function POST(request) {
...
+    } else {
+      players.teamB.push({ player_id, pname });
+    }
+
+    // run this right before match cutover (on the last item in data for a match)
+    if (i === matchData.length - 1 || matchData[i].match_id !== matchData[i + 1].match_id) {
+      Object.assign(rest, players);
+      rest.mdate = new Date(
+        rest.mdate.toISOString().slice(0, 10) + 'T00:00:00.000-05:00',
+      ).toDateString();
```
Comment (KyungjinJPark) [inline next-app/app/api/matches/route.js:54]: why we chop off timestamp?

File: next-app/app/components/navbar/navbar.js [hunk 9]
```diff
@@ -1,20 +1,29 @@
...
+    '/players': 'Players',
+  };
+
   return (
     <Navbar expand="lg" className="bg-body-tertiary">
       <Container>
         <Navbar.Brand href="/">
           <Image alt="" src="/logo.png" style={{ borderRadius: 7 }} width="48" height="48"></Image>
         </Navbar.Brand>
+        <h1 className="d-lg-none display-3">{page[usePathname()]}</h1>
```
Comment (KyungjinJPark) [inline next-app/app/components/navbar/navbar.js:21]: i think there may be some mobile / desktop breakpoint mismatches somewhere. when my width is 1150px, i dont get the page name in the page or the nav bar

File: next-app/app/matches/page.js [hunk 10]
```diff
@@ -68,25 +78,18 @@ function MatchData({ matches }) {
...
+  const reformat = matches.map((value) => {
+    const { teamA, teamB, ...rest } = value;
+    const a = value.teamA.map((a) => {
+      return a.pname;
+    });
+    const b = value.teamB.map((b) => {
+      return b.pname;
+    });
+    rest.teamA = a;
+    rest.teamB = b;
```
Comment (KyungjinJPark) [inline next-app/app/matches/page.js]: you could skip a step and set the properties directly (also, remove the `value.` before accessing the team since you already destructured it out)
```suggestion
    rest.teamA = teamA.map((a) => {
      return a.pname;
    });
    rest.teamB = teamB.map((b) => {
      return b.pname;
    });
```

File: next-app/app/page.js [hunk 11]
```diff
@@ -1,30 +1,127 @@
...
+    refetch,
+  } = useRequest('GET', '/api/schedule');
+  const [showToast, setShowToast] = useState(false);
+  const [dateRequested, setDateRequested] = useState(null);
 
-  if (loading) {
+  function scheduleDupe(date) {
+    let duplicate = false;
+
+    schedule.forEach((schedule) => {
```
Comment (KyungjinJPark) [inline next-app/app/page.js]: could look into `Array.prototype.some()` for this

File: next-app/app/page.js [hunk 12]
```diff
@@ -1,30 +1,127 @@
...
+
+      if (!response.ok) {
+        throw new Error(`Response status: ${response.status}`);
+      }
+
+      refetch();
+    } catch (error) {
+      console.error(error.message);
+    }
+  }
```
Comment (KyungjinJPark) [inline next-app/app/page.js:96]: line 24 to here 101, if none of this logic is used within this component `Home`, it may be best to move the logic into `WeatherData`

File: next-app/app/page.js [hunk 13]
```diff
@@ -1,30 +1,127 @@
...
-  if (error) {
+  if (weatherErr || scheduleErr) {
     return <Error />;
   }
 
   return (
     <div>
-      <h1 className="display-1 text-center p-auto m-auto pt-2">PikoWeatherer</h1>
+      <h1 className="display-1 text-center p-auto m-auto pt-2 title">PikoWeatherer</h1>
       <br />
```
Comment (KyungjinJPark) [inline next-app/app/page.js]: i know these `br`s arent part of this PR, but generally, we should try to avoid them and us HTML/CSS margin/padding features

File: next-app/app/page.js [hunk 14]
```diff
@@ -37,24 +134,25 @@ export default function Home() {
...
+function WeatherData({
+  weatherData,
+  addDate,
+  showToast,
+  setShowToast,
+  dateRequested,
+  dupe,
+  undoDate,
+}) {
+  const weekday = {
```
Comment (KyungjinJPark) [inline next-app/app/page.js]: all this static data can probably be taken out of components, bc it doesn't depend on any state

File: next-app/app/api/matches/route.js [hunk 15]
```diff
@@ -34,3 +34,28 @@ export async function POST(request) {
...
+  let players = { teamA: [], teamB: [] };
+  for (let i = 0; i < matchData.length; i++) {
+    const { pname, team, player_id, ...rest } = matchData[i];
+    if (team === 'a') {
+      players.teamA.push({ player_id, pname });
+    } else {
+      players.teamB.push({ player_id, pname });
+    }
+
+    // run this right before match cutover (on the last item in data for a match)
```
Comment (KyungjinJPark) [inline next-app/app/api/matches/route.js]: could look into `jsDoc` for this

File: next-app/app/players/page.js [hunk 16]
```diff
@@ -156,7 +166,7 @@ function calculatePlayerStats(players, matches) {
     if (player.matches < 5) player.titles.push('Sprout');
     if (player.matches > 30) player.titles.push('Veteran');
     if (player.matches > 100) player.titles.push('Legend');
-    if (player.matches > 5 && player.winrate >= 0.75) player.titles.push('Farmer');
+    if (player.winrate >= 0.75) player.titles.push('Farmer');
```
Comment (KyungjinJPark) [inline next-app/app/players/page.js]: i think we want to keep this match threshold

File: next-app/app/schedule/page.tsx [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+  if (scheduleError || matchesError) {
+    return <ErrorComponent />;
+  }
+
+  return (
+    <div>
+      <h1 className="display-1 text-center p-auto m-auto pt-2 title">Schedule</h1>
+      <br />
+      <ScheduleData schedule={scheduleDB} matches={matchesDB} unschedule={unschedule} />
+      {/* <ConfirmationModal></ConfirmationModal> */}
```
Comment (KyungjinJPark) [inline next-app/app/schedule/page.tsx]: 🙈🙊🙉

File: next-app/app/schedule/page.tsx [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  if (scheduleError || matchesError) {
+    return <ErrorComponent />;
+  }
+
+  return (
+    <div>
+      <h1 className="display-1 text-center p-auto m-auto pt-2 title">Schedule</h1>
+      <br />
```
Comment (KyungjinJPark) [inline next-app/app/schedule/page.tsx]: same comment above about brs

File: next-app/app/schedule/page.tsx [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+  } else {
+    return (
+      <div className="text-center">
+        <p>No scheduled dates found</p>
+      </div>
+    );
+  }
+}
+
+function reformatMatches(matches) {
```
Comment (KyungjinJPark) [inline next-app/app/schedule/page.tsx:103]: since this is used more than once, you may want to make a utility function here

File: next-app/app/page.js (hunk 12 above)
Comment (KyungjinJPark) [reply next-app/app/page.js:96]: however, i think the wierd behavior where the toast moves with the card has to do with the toast being a child of the card and not the greater page. some of this logic should stay here were you to fix this error

File: next-app/app/schedule/page.tsx [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+          })}
+        </tbody>
+      </table>
+    );
+  } else {
+    return <p>No match data found.</p>;
+  }
+}
+
+function ConfirmationModal({ unscheduleID, unscheduledDate }) {
```
Comment (KyungjinJPark) [inline next-app/app/schedule/page.tsx]: may want the name to be descriptive in a way that implies the button is a part of this component. that or take the button out of the component.

**Review by KyungjinJPark — CHANGES_REQUESTED** [review]

**Comment by KyungjinJPark** [comment]
you got some merge conflicts

File: next-app/app/players/page.js (hunk 16 above)
Comment (thomasp18) [reply next-app/app/players/page.js]: i didnt touch the players page other than the fix that i did. Alex must've updated after I created this branch

File: next-app/app/api/matches/route.js (hunk 8 above)
Comment (thomasp18) [reply next-app/app/api/matches/route.js:54]: it breaks the timezone bs

File: next-app/app/page.js (hunk 11 above)
Comment (thomasp18) [reply next-app/app/page.js]: nah

File: next-app/app/api/matches/route.js (hunk 15 above)
Comment (thomasp18) [reply next-app/app/api/matches/route.js]: 🤔

File: next-app/app/page.js (hunk 12 above)
Comment (thomasp18) [reply next-app/app/page.js:96]: MonkeyCuteGIF
will keep this in mind!

**Review by KyungjinJPark — APPROVED** [review]
